### PR TITLE
[FLINK-25379][connectors] Support limit push down in DATAGEN connector

### PR DESCRIPTION
**Pull request by ZhangChaoming**
## What is the purpose of the change

Support limit push down in DATAGEN connector.

## Brief change log

  - Support limit push down in DATAGEN connector.

## Verifying this change

Apply `LIMIT` in datagen table query, see `DataGeneratorConnectorITCase#testLimit`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / **don't know**)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
You can already specify the number of records for this connector per the documentation, see https://nightlies.apache.org/flink/flink-docs-stable/docs/connectors/table/datagen/

**Review by slinkydeveloper — COMMENTED** [review]
Nice change, can you add a simple test in `DataGeneratorConnectorITCase` that checks that using the `LIMIT` keyword instead of the `number-of-rows` option works fine?

**Comment by slinkydeveloper** [comment]
@MartijnVisser this PR adds support for the limit keyword, that is instead of defining `number-of-rows` per doc, you can also omit that option and instead per query set the limit like this `SELECT * FROM datagentable LIMIT 10`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:81]: Since this table is still defined with `number-of-rows = 5`, this test doesn't actually assert that the limit keyword does anything.

Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:81]: This table actually defined with `'number-of-rows' = '10'`.

Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:81]: Agree with @Airblader, please add a test with a new table definition that does not include `number-of-rows`, as it also tests that the validation phase doesn't fail because of missing `number-of-rows`

Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java]: OK

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:102]: Please use assertj assertions here, and convert the rest of this class to use assertj as well

Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java]: OK

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:103]: `assertThat(results).hasSize(5)`. Same above

**Comment by slinkydeveloper** [comment]
Thank you for the contribution @ZhangChaoming, just left another small comment and then it looks good!

**Comment by MartijnVisser** [comment]
I am wondering if LIMIT is the best option for the Datagen connector, given the special circumstances for this connector because it generates random rows every time. I think LIMIT implies that a user wants to limit the available (or in this case, generated) results. If I would apply a LIMIT in combination with a specific reading position for Kafka, I get the same results over and over. That's not the case for Datagen of course, because it generates random values.

I rather be consistent in what LIMIT does for all connectors and keep the current method of `number-of-rows` to change the connector from unbounded to bounded, so a user doesn't have to read and understand the documentation that LIMIT does something differently for Datagen compared to other connectors.

**Comment by Airblader** [comment]
The expectation that a query with `LIMIT` always returns the same results is only true for ordered output, which streaming queries don't support to begin with. I don't think the comparison to the Kafka connector is correct here; having to specify a fixed offset is an additional requirement (not mandated by the Kafka connector), and a similar thing could be done for datagen (specifying a seed for the RNG).

Let's keep in mind that the limit _is already_ being applied, by Flink itself outside the source. The pushdown just allows the source itself to aid with that. I don't see a great reason why the datagen connector should need to support this, but I certainly don't see harm in it either.

**Comment by MartijnVisser** [comment]
> The expectation that a query with `LIMIT` always returns the same results is only true for ordered output, which streaming queries don't support to begin with.

For me the question is what a user expects when its applying LIMIT. The fact we don't (yet) support it is less relevant in that context. Is there anything on LIMIT in the SQL standards? 

> I don't see a great reason why the datagen connector should need to support this, but I certainly don't see harm in it either.

The harm I see is that we have two options to achieve apparently the same result, which could also be conflicting with each other ("What output can I expect when I set `number-of-rows` to 10 and `LIMIT 5`")?

**Comment by sjwiesman** [comment]
@MartijnVisser 

A datagen table with a set number of rows is `BOUNDED` and can be run under batch execution (this is why we added it originally). `LIMIT` is technically an orthogonal concept, but I agree it is murky.

**Comment by Airblader** [comment]
> The harm I see is that we have two options to achieve apparently the same result

We already do, though, because `LIMIT` already works. It's just not pushed into the source.

**Comment by Airblader** [comment]
> The harm I see is that we have two options to achieve apparently the same result

But `LIMIT` does already work, it's just not pushed into the source (for this connector). For the user nothing really changes here(?)

> What output can I expect when I set `number-of-rows` to 10 and `LIMIT 5`

Why should it behave differently than for any other bounded source to which a limit is applied?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:96]: This test actually still doesn't test the limit pushdown support of this source as the test cannot distinguish that from the limit applied generically over all sources.

Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java]: Maybe we can inffer that from it's logical plan.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java]: OK

**Comment by MartijnVisser** [comment]
> But `LIMIT` does already work, it's just not pushed into the source (for this connector). For the user nothing really changes here(?)

If that's the case, then I agree that my argument is not applicable here. 

> Why should it behave differently than for any other bounded source to which a limit is applied?

It shouldn't, but Datagen itself behaves differently than any other bounded source because of its randomness.

**Comment by MartijnVisser** [comment]
@ZhangChaoming Please have a look first at the CI, because the build is currently not passing

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@MartijnVisser I can not find any CI progress for this PR.

**Comment by MartijnVisser** [comment]
@ZhangChaoming It always updates the one from the first comment, which you can find here https://github.com/apache/flink/pull/18151#issuecomment-997659194

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@MartijnVisser Thanks, check code style failed, I will fix it.

**Comment by ZhangChaoming** [comment]
@MartijnVisser Pipeline still failed, but it seems the error was not caused by my code.

**Comment by MartijnVisser** [comment]
@ZhangChaoming You should probably rebase your branch on the latest master to resolve those issues

**Comment by ZhangChaoming** [comment]
@MartijnVisser I think my code is safe, since the error was caused by testing `test_ci finegrained_resource_management`. And This branch has no conflict with the base branch.

**Comment by MartijnVisser** [comment]
@ZhangChaoming Still, I wouldn't merge it unless CI has fully passed. Note: you'll need a review from probably @slinkydeveloper before this could be merged

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@MartijnVisser The CI has passed, and could you please review this PR?

**Comment by MartijnVisser** [comment]
@ZhangChaoming I hope that @Airblader or @slinkydeveloper can take care of the review

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:101]: This exception is a bit hard to understand and relies on string comparison, can you at least just do the contains on the original explain string without the splitting?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:104]: no need to declare the `rows` variable, inline that

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:101]: OK

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java
Comment (ZhangChaoming) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/runtime/stream/table/DataGeneratorConnectorITCase.java:104]: OK

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@flinkbot run azure

**Comment by ZhangChaoming** [comment]
@MartijnVisser Could you please help review this ?

**Comment by MartijnVisser** [comment]
> slink

@slinkydeveloper Can you help with the review?

**Review by slinkydeveloper — APPROVED** [review]

**Comment by MartijnVisser** [comment]
@slinkydeveloper Thanks for the review and @ZhangChaoming thanks for the PR. I've merged it.

**Pull request merged**
